_posts/2024-04-30-search-indexing-rollover.adoc

```diff
@@ -0,0 +1,281 @@
+:imagesdir: /assets/images/posts/search-indexing-rollover
+:hibernate-search-docs-url: https://docs.jboss.org/hibernate/search/{hibernate-search-version}/reference/en-US/html_single/
+:quarkus-hibernate-search-docs-url: https://quarkus.io/guides/hibernate-search-orm-elasticsearch
+
+This is the first post in the series diving into the implementation details of the
+link:https://github.com/quarkusio/search.quarkus.io[application] backing the guide search of
+link:https://quarkus.io/guides/[quarkus.io].
+
+Does your application have full-text search capabilities and use Hibernate Search?
+Do you need to perform reindexing of your data while keeping your application running and producing search results?
+Look no further. In this post, we'll cover how you can approach this problem
+and solve it in practice with a few low-level APIs provided by Hibernate Search.
+
+The approach suggested in this post is based on the fact that Hibernate Search is using
+link:{hibernate-search-docs-url}#backend-elasticsearch-indexlayout[aliased indexes],
+and communicates with the actual index through a read/write alias, depending on the operation it needs to perform.
+For example, a search operation will be routed to a read index alias,
+while the indexing operation will be sent to a write index alias.
+
+image::initial-app.png[]
+
+NOTE: This approach is implemented and successfully used in our Quarkus application that backs the guides'
+search of https://quarkus.io/guides/[quarkus.io/guides/].
+You can see the complete implementation here:
+link:https://github.com/quarkusio/search.quarkus.io/blob/d956b6a1341d8693fa1d6b7881f3840f48bdaacd/src/main/java/io/quarkus/search/app/indexing/Rollover.java#L44-L331[rollover implementation]
+and link:https://github.com/quarkusio/search.quarkus.io/blob/d956b6a1341d8693fa1d6b7881f3840f48bdaacd/src/main/java/io/quarkus/search/app/indexing/IndexingService.java#L226-L244[rollover usage].
+
+Applications using Hibernate Search can keep their search indexes up-to-date by updating the index gradually,
+as the data on which the index documents are based is modified, providing a near real-time index synchronisation.
+On the other hand, if the search requirements allow for a delay in synchronisation
+or the data is updated only at certain times of day, the option of mass indexing can effectively keep the indexes up-to-date.
+The link:{hibernate-search-docs-url}[Hibernate Search documentation] provides more information about these approaches
+and other Hibernate Search capabilities.
+
+The application discussed in this post is using the mass indexing approach.
+This means that at certain events, e.g., when a new version of the application is deployed or a scheduled time is reached,
+the application has to process the documentation guides and create search index documents from them.
+
+Now, since we would want to keep our application providing results to any search operations and add/update documents to the indexes
```

Comment:
```suggestion
Now, since we want our application to keep  providing results to any search requests while we add/update documents to the indexes,
```